tests/two_d_array_int/two_d_array_int.c

```diff
@@ -0,0 +1,251 @@
+    CHAOS_TEST_ACTION_GET_ROW \
+
+MU_DEFINE_ENUM(CHAOS_TEST_ACTION, CHAOS_TEST_ACTION_VALUES)
+MU_DEFINE_ENUM_STRINGS(CHAOS_TEST_ACTION, CHAOS_TEST_ACTION_VALUES)
+
+BEGIN_TEST_SUITE(TEST_SUITE_NAME_FROM_CMAKE)
+
+TEST_SUITE_INITIALIZE(suite_init)
+{
+}
+
+TEST_SUITE_CLEANUP(suite_cleanup)
+{
+}
+
+TEST_FUNCTION_INITIALIZE(method_init)
+{
+}
+
+TEST_FUNCTION_CLEANUP(method_cleanup)
+{
+}
+
+TEST_FUNCTION(TWO_D_ARRAY_CREATE_create_succeeds)
+{
+    //arrange
+
+    //act
+    TWO_D_ARRAY(TEST_THANDLE) tdarr = TWO_D_ARRAY_CREATE(TEST_THANDLE)(5, 5);
+
+    //assert
+    ASSERT_IS_NOT_NULL(tdarr);
+
+    //cleanup
+    TWO_D_ARRAY_ASSIGN(TEST_THANDLE)(&tdarr, NULL);
+}
+
+#define ROW_NUM 1000
+
+TEST_FUNCTION(TWO_D_ARRAY_create_and_allocate_succeeds)
+{
+    //arrange
+    TWO_D_ARRAY(TEST_THANDLE) tdarr = TWO_D_ARRAY_CREATE(TEST_THANDLE)(ROW_NUM, 5);
+    ASSERT_IS_NOT_NULL(tdarr);
+
+    //act
+    for (uint32_t i = 0; i < ROW_NUM; i++)
+    {
+        TWO_D_ARRAY_ALLOCATE_NEW_ROW(TEST_THANDLE)(tdarr, i);
+    }
+
+    //assert
+    for (uint32_t i = 0; i < ROW_NUM; i++)
+    {
+        ASSERT_IS_NOT_NULL(tdarr->row_arrays[i]);
+    }
+
+    //cleanup
+    TWO_D_ARRAY_ASSIGN(TEST_THANDLE)(&tdarr, NULL);
+}
+
+TEST_FUNCTION(TWO_D_ARRAY_allocate_row_and_free_row_succeeds)
+{
+    //arrange
+    TWO_D_ARRAY(TEST_THANDLE) tdarr = TWO_D_ARRAY_CREATE(TEST_THANDLE)(ROW_NUM, 5);
+    ASSERT_IS_NOT_NULL(tdarr);
+    for (uint32_t i = 0; i < ROW_NUM; i++)
+    {
+        TWO_D_ARRAY_ALLOCATE_NEW_ROW(TEST_THANDLE)(tdarr, i);
+    }
+
+    //act
+    for (uint32_t i = 0; i < ROW_NUM; i++)
+    {
+        TWO_D_ARRAY_FREE_ROW(TEST_THANDLE)(tdarr, i);
+    }
+
+    //assert
+    for (uint32_t i = 0; i < ROW_NUM; i++)
+    {
+        ASSERT_IS_NULL(tdarr->row_arrays[i]);
+    }
+
+    //cleanup
+    TWO_D_ARRAY_ASSIGN(TEST_THANDLE)(&tdarr, NULL);
+}
+
+TEST_FUNCTION(TWO_D_ARRAY_allocate_a_row_already_been_allocated)
+{
+    //arrange
+    TWO_D_ARRAY(TEST_THANDLE) tdarr = TWO_D_ARRAY_CREATE(TEST_THANDLE)(ROW_NUM, 5);
+    ASSERT_IS_NOT_NULL(tdarr);
+    int result;
+    result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(TEST_THANDLE)(tdarr, 0);
+    ASSERT_ARE_EQUAL(int, result, 0);
+
+    //act
+    result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(TEST_THANDLE)(tdarr, 0);
+
+    //assert
+    ASSERT_ARE_NOT_EQUAL(int, result, 0);
+
+    //cleanup
+    TWO_D_ARRAY_ASSIGN(TEST_THANDLE)(&tdarr, NULL);
+}
+
+TEST_FUNCTION(TWO_D_ARRAY_GET_ROW_succeeds)
+{
+    //arrange
+    TEST_THANDLE* row_result;
+    TWO_D_ARRAY(TEST_THANDLE) tdarr = TWO_D_ARRAY_CREATE(TEST_THANDLE)(5, 5);
+    int add_row_result = TWO_D_ARRAY_ALLOCATE_NEW_ROW(TEST_THANDLE)(tdarr, 0);
+
+    ASSERT_IS_NOT_NULL(tdarr);
+    ASSERT_ARE_EQUAL(int, add_row_result, 0);
+    ASSERT_IS_NOT_NULL(tdarr->row_arrays[0]);
+    for (unsigned int i = 0; i < 5; i++)
+    {
+        tdarr->row_arrays[0][i].a = i;
+    }
+
+    //act
+    row_result = TWO_D_ARRAY_GET_ROW(TEST_THANDLE)(tdarr, 0);
```

Comment:
>row_result = TWO_D_ARRAY_GET_ROW(TEST_THANDLE)(tdarr, 0); [](http://example.com/codeflow?start=4&length=57)

I'm curious about the expected usage here. A few tests up we directly access `tdarr->row_arrays`, but now I see we have a function to do it (which seems to do the bounds checking. Should line 115 and others above be calling this instead?

Do we want to "hide" the `row_arrays` field from "users"?

Reply:
yes, want to hide row_arrays, changed to use TWO_D_ARRAY_GET_ROW in other places